Improve question documentation (#11830)
PrairieLearn/PrairieLearn: docs/question/server.md

# `server.py`

The `server.py` file for each question creates randomized question variants by generating random parameters and the corresponding correct answers.

## Guide

To explain how this works, we will use a simple example of a question that asks the student to double a number. The question will look like this:

<!-- prettier-ignore -->
```html title="question.html"
<pl-question-panel>
  If $x = {{params.x}}$, what is $y$ if $y$ is double $x$?
</pl-question-panel>
<pl-number-input answers-name="y" label="$y =$"></pl-number-input>
<pl-submission-panel>
  {{feedback.y}}
</pl-submission-panel>
```

> More details about the `{{params.x}}` and `{{feedback.y}}` syntax can be found in the [`question.html` documentation](./html.md).

Reviewer: ```suggestion
More details about the `{{params.x}}` and `{{feedback.y}}` Mustache syntax can be found in the [`question.html` documentation](./html.md).
```

### Step 1: `generate`

First, the `generate` function is called to generate random parameters for the variant, and the correct answers. It should set `data["params"]` with the parameters for the question, and `data["correct_answers"]` with the correct answers. The parameters can then be used in the `question.html` file by using `{{params.NAME}}`.

```python title="server.py"
import random

def generate(data):
    # Generate random parameters for the question and store them in the data["params"] dict:
    data["params"]["x"] = random.randint(5, 10)

    # Also compute the correct answer (if there is one) and store in the data["correct_answers"] dict:
    data["correct_answers"]["y"] = 2 * data["params"]["x"]
```

!!! info

>     In general, each function of the question generation process runs *after* all elements in the question. For example, the `generate()` function in `server.py` runs after all elements have run their `generate()` functions. This is important to remember when using the `data` dictionary, as it will contain the results *after* the elements have finished processing.

Reviewer 2: I think the placement of this info box is a bit misleading, since none of the provided elements use a `generate` function. I'd suggest moving this either to the next step or before all steps. In the latter case, `parse()` and `grade()` might be better examples since they are often combined (element+question) in cases where custom grading is used.

Author: I moved the info box to `parse` in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9.

### Step 2: `prepare`

Next, the `prepare` function is called after all elements (e.g. `<pl-number-input>`) have run `generate()`. This is typically done to do any sort of final post-processing, but is not commonly used.

### Step 3: `render`

> Next, the `render(data)` function is called to render the question. You can use this function to override how the question is rendered. This is typically only used for more advanced questions.

Reviewer 2: Is this enough to satisfy #7339? Or should this have an example?

Author: I added some more text in [8c7ac2d](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/8c7ac2d708e1c9e582772b4f4d95b962e0e2d6a9). If I add an example, I would want it in a collapsible section. If you want to provide an example, that would be excellent, but otherwise, maybe we can save that for a follow up PR.

Reviewer 3: Yeah, IMO we should save net-new content like this for another PR. I don't even really know how to talk about `server.py#render`. What's it good for? When do we suggest using it? What's a compelling example of how it can be used?

Reviewer 2: The only situation where I used render in the past was to manually "fix" the rendering of an answer panel after the variants had already been created. In particular, when I first created the question the answer panel was empty, but later on I decided to provide a suggested answer (code). The problem was that one of the parts of the code was a math expression that depended on the generated params, and the expression was complex enough that using mustache was not enough. A proper way to handle this for usual cases would be to generate the expression in params and add it to the answer panel, but since I couldn't run generate again I had to improvise. [Here's the code](https://github.com/PrairieLearn/pl-yorku-digt2103/blob/839849e409505de9c8b6473cc3643f0a928c7ff2/questions/exam2_integer_sequence/server.py#L65-L72) for those with access.

That said, I agree that this function would not be used very often. I think just describing it should be enough, with a comment that it is intended for advanced users.

Author: > Next, the `render(data, html)` function is called to render the question. You can use this function to override how the question is rendered. The render function expects two parameters, `data` and `html`, and should return a string of HTML. The HTML after the mustache template has been rendered is available through the `html` parameter. This is typically only used for more advanced questions, and we won't need it for this example.

is the current text. @jonatanschroeder how does that look?

### Step 4: `parse`

> When a student submits their answer, the `parse` function is called to parse the submitted answers after the individual elements have parsed them. This function can be used to display more-specific format errors than the input elements or to parse the input differently. When our parse function runs, `<pl-number-input>` will have already parsed the submitted value as an integer, and display an error to the student if it was invalid. For this question, we will allow the student to only submit positive integers, so we will check for that and set a format error with `data["format_errors"]` if it is negative.

Reviewer 2: I don't know if it may be a good idea to also expand the explanation (even if not applicable to the example) that parse can also be used parse the input in a question-specific way. In other words, parse is not just to check for errors, it can also be used to interpret the input.

Author: I expanded the example in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9 showing how you could modify the submitted answer in `parse`.

- `data["raw_submitted_answers"][NAME]` is the exact raw answer submitted by the student.
- `data["submitted_answers"][NAME]` is the answer parsed by elements (e.g. strings converted to numbers).
- `data["format_errors"][NAME]` is the format error message for a named answer.

> If there are format errors then the submission is "invalid" and is not graded. To provide feedback but keep the submission "valid", `data["feedback"][NAME]` can be used instead.

Reviewer: ```suggestion
If there are format errors, then the submission is "invalid" and is not graded. To provide feedback but keep the submission "valid", `data["feedback"][NAME]` can be used instead.
```

```python title="server.py"
def parse(data):
    # Reject negative numbers for "y" if we don't already have a format error
    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
        data["format_errors"]["y"] = "Negative numbers are not allowed"
```

### Step 5: `grade`

Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:

- Setting the score and feedback for each named answer in `data["partial_scores"]`.
- Setting the total score for the question in `data["score"]`.
- Setting the overall feedback for the question in `data["feedback"]`.

This function only runs if `parse()` did not produce format errors, so we can assume all data is valid. All elements will have already graded their answers (if any) before this point.

> - `data["partial_scores"][NAME]["score"]` is the individual scores for the named answers (0 to 1).

Reviewer 2: I think you should also point out `data["partial_scores"][NAME]["feedback"]`, which is the element-specific feedback (as opposed to the question-wide feedback).

Author: Pointed this out in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9. I also added a note that instructors shouldn't use this field for giving feedback.

Reviewer 2: I don't know if I agree with that note. We use the partial scores feedback in our own example course questions.

https://github.com/PrairieLearn/PrairieLearn/blob/b1929d0341d01e6ed1d94bfb3e1293a89a435d9a/exampleCourse/questions/demo/annotated/LectureVelocity/4-KnowledgeTest/server.py#L90-L102

One issue with `data["feedback"][NAME]` is that it is not shown by default, and instructors need to explicitly show it using `{{feedback.NAME}}`. The partial scores feedback can be shown by specific elements in an element-specific way.

https://github.com/PrairieLearn/PrairieLearn/blob/b1929d0341d01e6ed1d94bfb3e1293a89a435d9a/apps/prairielearn/elements/pl-number-input/pl-number-input.mustache#L132-L136

Author: Ah I wasn't aware of that. That is a very important detail!

Author: Updated in 21cca4540216bd25e2fa512cf2f8f4a3e2f34a7f.

Author: @jonatanschroeder I updated the feedback section further in f1cc70d4a6303cb66a9ad081fdc705ef1daa184e.

- `data["score"]` is the total score for the question (0 to 1).
- `data["feedback"][NAME]` is the feedback for each named answer.

It is recommended that you give additional feedback to the student as they make progress towards the solution, and reward this progress with partial credit.

If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback from the element that graded it. If the `grade` function _is_ defined, the data you receive has already been graded by the elements. You should ensure you only award partial credit if the answer isn't correct, otherwise you might give partial credit for a correct answer.

> You can set `data["format_errors"][NAME]` to mark the submission as invalid. This will cause the question to not use up one of the student's attempts on assessments. You are encouraged, though, to do any checks for invalid data that can be done in `parse(data)` there instead, since that method is also called when the student hits "Save only", in manually graded questions, or in assessments without real-time grading.

Reviewer: ```suggestion
You can set `data["format_errors"][NAME]` to mark the submission as invalid. This will cause the question to not use up one of the student's attempts on assessments. However, you are encouraged to do as many checks for invalid data as possible in `parse` instead of `grade`; the `parse` function is called when the student hits "Save only", in manually graded questions, and in assessments without real-time grading.
```

```python title="server.py"
import math

def grade(data):
    # Give half points for incorrect answers larger than "x", only if not already correct.
    # Use math.isclose to avoid possible floating point errors.
    y_is_correct = math.isclose(data["partial_scores"]["y"]["score"], 1.0)
    if not y_is_correct and data["submitted_answers"]["y"] > data["params"]["x"]:
        data["partial_scores"]["y"]["score"] = 0.5

>         data["score"] = set_weighted_score_data(data)

Reviewer 2: `set_weighted_score_data()` updates `data["score"]` directly, and does not return anything. You also need to import it.

Author: Fixed in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9.

# `server.py`

The `server.py` file for each question creates randomized question variants by generating random parameters and the corresponding correct answers.

## Guide

To explain how this works, we will use a simple example of a question that asks the student to double a number. The question will look like this:

<!-- prettier-ignore -->
```html title="question.html"

> <pl-question-panel>
>     If $x = {{params.x}}$, what is $y$ if $y$ is double $x$?
> </pl-question-panel>
> <pl-number-input answers-name="y" label="$y =$"></pl-number-input>
> <pl-submission-panel>
>     {{feedback.y}}
> </pl-submission-panel>

Reviewer 3: ```suggestion
<pl-question-panel>
  If $x = {{params.x}}$, what is $y$ if $y$ is double $x$?
</pl-question-panel>
<pl-number-input answers-name="y" label="$y =$"></pl-number-input>
<pl-submission-panel>
  {{feedback.y}}
</pl-submission-panel>
```

Author: 8de961fa8bacdf280e06dd8f5e73d124836d5b15

```

More details about the `{{params.x}}` and `{{feedback.y}}` syntax can be found in the [`question.html` documentation](./html.md).

### Step 1: `generate`

First, the `generate` function is called to generate random parameters for the variant, and the correct answers. It should set `data["params"]` with the parameters for the question, and `data["correct_answers"]` with the correct answers. The parameters can then be used in the `question.html` file by using `{{params.NAME}}`.

```python title="server.py"

> import random

Reviewer 3: Nit:

```suggestion
import random

```

Author: 8de961fa8bacdf280e06dd8f5e73d124836d5b15

```python title="server.py"
def parse(data):
    # check we don't already have a format error, and if not, check if the answer is negative
    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
        data["format_errors"]["y"] = "Negative numbers are not allowed"
```

### Step 5: `grade`

Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:

- Setting the score and feedback for each named answer in `data["partial_scores"]`.
- Setting the total score for the question in `data["score"]`.
- Setting the overall feedback for the question in `data["feedback"]`.

If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback.

> It is recommended that you give additional feedback to the student beyond right and wrong, as well as partial credit. If the function is defined, the data you receive has already been graded by the elements. You should make sure that you don't override a correct answer by only giving partial credit.

Reviewer 3: I'm not 100% sure what the last sentence here is supposed to mean?

Author: If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback. If the function _is_ defined, the data you receive has already been graded by the elements. You should ensure you don't lower the score that was already given by the elements. In the example below, we accomplish this using the `marked_as_incorrect` variable, which is set to `True` only if all elements (in this case, there is only one) have marked the named answers as incorrect. In this case, it is safe to award partial credit.

Reviewer 3: Continuing conversation here: https://github.com/PrairieLearn/PrairieLearn/pull/11830/files#r2054784983

You can also set `data["format_errors"]` to mark the submission as invalid. This will cause the question to not use up one of the student's attempts on assessments.

```python title="server.py"

> import math

Reviewer 3: Nit:

```suggestion
import math

```

Author: 8de961fa8bacdf280e06dd8f5e73d124836d5b15

def grade(data):
    marked_as_incorrect = math.isclose(data["score"], 0.0)
    if marked_as_incorrect and data["submitted_answers"]["y"] > data["params"]["x"]:
        data["partial_scores"]["y"]["score"] = 0.5
        data["score"] = 0.5
        data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
```

### Complete example

> The finished, complete `question.html` and `server.py` example looks like:

Reviewer 3: This example both looks and behaves different than what was built step-by-step above.

Author: Which one would you prefer? I did this to reduce complexity in the guide, but the complete example may be more realistic.

Reviewer 3: The main difference between this and the complete example is the randomized operation. I'd be happy to remove that in favor or just having a random parameter and correct answer.

The full example also contains a lot of inline documentation as comments. Would it be feasible to remove some of those (if they've already been discussed elsewhere) or pull them out into normal prose?

```python title="server.py"
def parse(data):
    # check we don't already have a format error, and if not, check if the answer is negative
    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
        data["format_errors"]["y"] = "Negative numbers are not allowed"
```

### Step 5: `grade`

Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:

- Setting the score and feedback for each named answer in `data["partial_scores"]`.
- Setting the total score for the question in `data["score"]`.
- Setting the overall feedback for the question in `data["feedback"]`.

It is recommended that you give additional feedback to the student as they make progress towards the solution, and reward this progress with partial credit.

> If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback. If the function _is_ defined, the data you receive has already been graded by the elements. You should ensure you don't lower the score that was already given by the elements. In the example below, we accomplish this using the `marked_as_incorrect` variable, which is set to `True` only if all elements (in this case, there is only one) have marked the named answers as incorrect. In this case, it is safe to award partial credit.

Author: I am trying to describe the behavior the example has, with `marked_as_incorrect`, and how we only award partial credit if the answer wasn't correct. I'm trying to say generally, a check like this is a good idea to avoid lowering a student score.

> If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback. If the function _is_ defined, the data you receive has already been graded by the elements. You should ensure you don't lower the score that was already given by the elements. In the example below, we accomplish this using the `marked_as_incorrect` variable, which is set to `True` only if all elements (in this case, there is only one) have marked the named answers as incorrect. In this case, it is safe to award partial credit.

Author: Maybe the example should be changed to `!= 1.0`, which I think is closer to the behavior you are describing.

> If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback. If the function _is_ defined, the data you receive has already been graded by the elements. You should ensure you don't lower the score that was already given by the elements. In the example below, we accomplish this using the `marked_as_incorrect` variable, which is set to `True` only if all elements (in this case, there is only one) have marked the named answers as incorrect. In this case, it is safe to award partial credit.

Reviewer 3: Ah, sure, I guess the situation here is really about a case where element code and `server.py` collaborate to produce a final score?

We could make this example more generalizable by checking `data["partial_scores"]["y"]["score"]` instead of `data["score"]`.

Let's leave this conversation unresolved so that @mwest1066 can take a look at this, he might have some ideas about how to better describe this.

> If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback. If the function _is_ defined, the data you receive has already been graded by the elements. You should ensure you don't lower the score that was already given by the elements. In the example below, we accomplish this using the `marked_as_incorrect` variable, which is set to `True` only if all elements (in this case, there is only one) have marked the named answers as incorrect. In this case, it is safe to award partial credit.

Author: Here is another pass:
> If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback from the element that graded it. If the `grade` function _is_ defined, the data you receive has already been graded by the elements. You should ensure you only award partial credit if the answer isn't correct, otherwise you might give partial credit for a correct answer.

> If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback. If the function _is_ defined, the data you receive has already been graded by the elements. You should ensure you don't lower the score that was already given by the elements. In the example below, we accomplish this using the `marked_as_incorrect` variable, which is set to `True` only if all elements (in this case, there is only one) have marked the named answers as incorrect. In this case, it is safe to award partial credit.

Author: > We could make this example more generalizable by checking data["partial_scores"]["y"]["score"] instead of data["score"].

This seems like the right approach.

> If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback. If the function _is_ defined, the data you receive has already been graded by the elements. You should ensure you don't lower the score that was already given by the elements. In the example below, we accomplish this using the `marked_as_incorrect` variable, which is set to `True` only if all elements (in this case, there is only one) have marked the named answers as incorrect. In this case, it is safe to award partial credit.

Author: Changed in 56ad65ab7b76953feff2ecafb26cec37104c9838.

- `data["raw_submitted_answers"][NAME]` is the exact raw answer submitted by the student.
- `data["submitted_answers"][NAME]` is the answer parsed by elements (e.g. strings converted to numbers).
- `data["format_errors"][NAME]` is the format error message for a named answer.

If there are format errors, then the submission is "invalid" and is not graded. To provide feedback but keep the submission "valid", `data["feedback"][NAME]` can be used instead.

```python title="server.py"
def parse(data):
    # Reject negative numbers for "y" if we don't already have a format error
    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
        data["format_errors"]["y"] = "Negative numbers are not allowed"
```

> We could have also modified the submitted answers directly, if we need to process the inputs further. For example, if we wanted to ensure that the answer is always positive, we could do this:

Reviewer 2: ```suggestion
We could have also modified the submitted answers directly, if we need to process the inputs further. For example, if we wanted to accept negative answers as equivalent to their absolute value, we could do this instead:
```

```python title="server.py"
def parse(data):
    # Ensure that "y" is positive
    data["submitted_answers"]["y"] = abs(data["submitted_answers"]["y"])
```

!!! info

    In general, each function of the question generation process runs *after* all elements in the question. For example, the `parse()` function in `server.py` runs after all elements have run their `parse()` functions. This is important to remember when using the `data` dictionary, as it will contain the results *after* the elements have finished processing.

> Although questions with custom grading usually don't use the grading functions from individual elements, it is _highly_ recommended that built-in elements are used for student input, as these elements include helpful parsing and feedback by default. Parsed student answers are present in the `data["submitted_answers"]` dictionary.

Reviewer 2: This phrasing is too strict.

```suggestion
Although questions with custom grading may not rely on the full grading functionality from individual elements, it is _highly_ recommended that built-in elements are used for student input, as these elements include helpful parsing and feedback by default. Parsed student answers are present in the `data["submitted_answers"]` dictionary.
```

    [PrairieLearn elements documentation](../elements.md) for more detailed discussion related to specific elements.

## Step 5: `grade`

Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:

- Setting the score and feedback for each named answer in `data["partial_scores"]`.
- Setting the total score for the question in `data["score"]`.
- Setting the overall feedback for the question in `data["feedback"]`.

This function only runs if `parse()` did not produce format errors, so we can assume all data is valid. All elements will have already graded their answers (if any) before this point.

- `data["partial_scores"][NAME]["score"]` is the individual scores for the named answers (0 to 1).
- `data["partial_scores"][NAME]["feedback"]` is the feedback for each named answer from the element that graded it.
- `data["score"]` is the total score for the question (0 to 1).
- `data["feedback"][NAME]` is the overall question feedback for each named answer.

It is recommended that you give additional feedback to the student as they make progress towards the solution, and reward this progress with partial credit.

> If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback from the element that graded it. If the `grade` function _is_ defined, the data you receive has already been graded by the elements. You should ensure you only award partial credit if the answer isn't correct, otherwise you might give partial credit for a correct answer. In the snipped below, we update `data["score"]` using the [`set_weighted_score_data`][prairielearn.question_utils.set_weighted_score_data] utility.

Reviewer 2: This last statement probably deserves a paragraph of its own, explaining why it's necessary (the grade function has the option of having the combined score of the question be computed from the partial scores or be overridden with a custom total) and that it's a 0-1 score.

```python title="server.py"
import math
import prairielearn as pl

def grade(data):
    # Give half points for incorrect answers larger than "x", only if not already correct.
    # Use math.isclose to avoid possible floating point errors.
    y_is_correct = math.isclose(data["partial_scores"]["y"]["score"], 1.0)
    if not y_is_correct and data["submitted_answers"]["y"] > data["params"]["x"]:
        data["partial_scores"]["y"]["score"] = 0.5
        pl.set_weighted_score_data(data)
        data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
```

### Providing feedback

To set custom feedback, the grading function should set the corresponding entry in the `data["feedback"]` dictionary. These feedback entries are passed in when rendering the `question.html`, which can be accessed by using the mustache prefix `{{feedback.}}`. See the [above example](#complete-example) or [this demo question](https://github.com/PrairieLearn/PrairieLearn/tree/master/exampleCourse/questions/demo/custom/gradeFunction) for examples of this.

> Some elements provide feedback through `data["partial_scores"][NAME]["feedback"]`, which you can override in the grading function. This field is a string, except for the `<pl-drawing>` and `<pl-checkbox>` elements. The feedback provided here will be attached to the student's answer for that element, which can make it easier for students to interpret the feedback they received in longer questions. Since not elements provide feedback this way, and use different data structures to represent the feedback they give, we recommend using `data["feedback"][NAME]` to provide question-specific feedback instead.

Reviewer 2: ```suggestion
Some elements provide feedback through `data["partial_scores"][NAME]["feedback"]`, which you can override in the grading function. This field is often a string, though some elements such as `<pl-drawing>` and `<pl-checkbox>` use different types. The feedback provided here will be attached to the student's answer for that element, which can make it easier for students to interpret the feedback they received in longer questions. Since not elements provide feedback this way, and use different data structures to represent the feedback they give, we recommend using `data["feedback"][NAME]` to provide question-specific feedback instead.
```

Author: This is just future proofing? I would prefer to give definites.

Reviewer 2: It's partly future proof, but also to cover course-level custom elements.

Reviewer 2: I'd prefer to phrase it in a way that ensures the instructor checks what's appropriate instead of relying on this to update the feedback blindly.

Author: How about a qualifier like `Some [official] elements`, so that it's clear we aren't talking about course-level custom elements?

Reviewer 2: I'd still prefer to have it clear that there is no implied consistency and that question writers should confirm before assuming.

Author: I think this ties into what we are discussing on the slack. about the standardization of the `partial_scores` interface.

Author: I'm fine making this change, but it would be nice to talk about this in #dev-core a little more first.

Author: Since there wasn't much discussion, I'll add the change you suggested, and maybe we can revisit this down the road.

Reviewer 4: I agree with @jonatanschroeder 's change here. 

Additionaly, with regards to this part:

> we recommend using `data["feedback"][NAME]` to provide question-specific feedback instead.

do we specify somewhere that they should insert that into the HTML themselves if they are going to do that?

Author: We do. https://prairielearn--11830.org.readthedocs.build/en/11830/question/server/#providing-feedback
![image](https://github.com/user-attachments/assets/55960e9e-8ef5-419f-afee-da3285f228bb)

!!! tip "Answer-specific feedback"

    If you want to provide feedback on a specific answer that doesn't use/support `data["partial_scores"][NAME]["feedback"]`,
    you could render the feedback directly after the element in `question.html`:

    <!-- prettier-ignore -->
    ```html title="question.html"
    <pl-string-input ...></pl-string-input>
    {{#feedback.value1}}<div>Feedback: {{.}}</div>{{/feedback.value1}}
    ```

> Overall question feedback in `data["feedback"][NAME]` needs be rendered explicity in the `question.html` template using `{{feedback.NAME}}`. Feedback given in `data["partial_scores"][NAME]["feedback"]` will be rendered automatically by the elements, assuming the element is visible.

Reviewer 2: ```suggestion
Overall question feedback in `data["feedback"][NAME]` needs be rendered explicity in the `question.html` template using `{{feedback.NAME}}`. Feedback given in `data["partial_scores"][NAME]["feedback"]`, if supported by the element, will be rendered automatically by the elements, assuming the element is visible.
```

Author: Is this redundant with the paragraph above? "Some elements provide feedback through..." and "Since not elements provide feedback this way,"

Reviewer 2: It's not redundant, it's consistent. As written it implies the feedback is the norm, not a limited feature.

Author: Makes sense, I had intended for a different interpretation but I didn't consider that one. You change was applied in [70ac407](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/70ac4076901e51062960fa1123b5d56f95a12a4b).

# `server.py`

The `server.py` file for each question creates randomized question variants by generating random parameters and the corresponding correct answers.

The following sections build a simple example of a question that asks the student to double a number. The question will look like this:

<!-- prettier-ignore -->
```html title="question.html"
<pl-question-panel>
  If $x = {{params.x}}$, what is $y$ if $y$ is double $x$?
</pl-question-panel>

> <pl-number-input answers-name="y" label="$y =$"></pl-number-input>

Author: Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.

    ```python title="server.py"
    from faker import Faker
    fake = Faker()

    fake.name()
    # 'Lucy Cechtelar'
    ```

### Parameter generation

For generated floating point answers, it's important to use consistent rounding when displaying numbers to students _and_ when computing the correct answer. Displaying rounded numbers to students but using the unrounded version for grading can lead to unexpected results.

=== "Good"

    ```python title="server.py"
    def generate(data):
        # Rounds numbers at the beginning
        a = np.round(33.33337, 2)
        b = np.round(33.33333, 2)
        data["params"]["a_for_student"] = f'{a:.2f}'
        data["params"]["b_for_student"] = f'{b:.2f}'
        data["correct_answers"]["c"] = a - b
    ```

=== "Bad"

    ```python title="server.py"
    def generate(data):
        a = 33.33337
        b = 33.33333
        data["params"]["a_for_student"] = f'{a:.2f}'
        data["params"]["b_for_student"] = f'{a:.2f}'
        # Correct answer is computed with full precision,
        # but the parameters displayed to students are rounded.
        data["correct_answers"]["c"] = a - b
    ```

## Step 2: `prepare`

> Next, the `prepare` function is called after all elements (e.g. `<pl-number-input>`) have run `prepare()`. This is typically done to do any sort of final post-processing, but is not commonly used.

Author: Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.

## Step 3: `render`

> Next, the `render(data, html)` function is called to render the question. You can use this function to override how the question is rendered. The render function expects two parameters, `data` and `html`, and should return a string of HTML. The HTML after the mustache template has been rendered is available through the `html` parameter. This is typically only used for more advanced questions, and we won't need it for this example.

Author: Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.

The values typically read or modified by a `parse` function are:

- `data["raw_submitted_answers"][NAME]` - the exact raw answer submitted by the student.
- `data["submitted_answers"][NAME]` - the answer parsed by elements (e.g. strings converted to numbers).
- `data["format_errors"][NAME]` - the format error message for a named answer.

If there are format errors, then the submission is "invalid" and is not graded. To provide feedback but keep the submission "valid", `data["feedback"][NAME]` can be used instead.

```python title="server.py"
def parse(data):
    # Reject negative numbers for "y" if we don't already have a format error
    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
        data["format_errors"]["y"] = "Negative numbers are not allowed"
```

We could have also modified the submitted answers directly, if we need to process the inputs further. For example, if we wanted to accept negative answers as equivalent to their absolute value, we could do this instead:

```python title="server.py"
def parse(data):
    # Ensure that "y" is positive

>     data["submitted_answers"]["y"] = abs(data["submitted_answers"]["y"])

Author: This is the type of edge case that I'm hoping #11856 would fix.

>     data["submitted_answers"]["y"] = abs(data["submitted_answers"]["y"])

Author: Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.

>     data["submitted_answers"]["y"] = abs(data["submitted_answers"]["y"])

Author: and 6baf6d2764c89bf89dc0530bbc266c77c679f5f9.

| `prepare()`  | :white_check_mark: | `answers_names`, `correct_answers`, `params`                                                             | Final question preparation after element code has run. Can modify data as necessary. Modify the `data` dictionary in-place.                                                                                                                                                   |
| `render()`   | :x:                | N/A. Returns `html` as a string                                                                          | Render the HTML for one panel and return it as a string.                                                                                                                                                                                                                      |

> | `parse()`    | :white_check_mark: | `format_errors`, `submitted_answers`, `correct_answers`, `feedback`                                      | Parse the `data["submitted_answers"][var]` data entered by the student, modifying this variable. Modify the `data` dictionary in-place.                                                                                                                                       |

Reviewer 2: All other entries have the keys in alphabetical order.
```suggestion
| `parse()`    | :white_check_mark: | `correct_answers`, `feedback`, `format_errors`, `submitted_answers`                                      | Parse the `data["submitted_answers"][var]` data entered by the student, modifying this variable. Modify the `data` dictionary in-place.                                                                                                                                       |
```

Author: Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.

| `grade()`    | :white_check_mark: | `correct_answers`, `feedback`, `format_errors`, `params`, `partial_scores`, `score`, `submitted_answers` | Grade `data["submitted_answers"][var]` to determine a score. Store the score and any feedback in `data["partial_scores"][var]["score"]` and `data["partial_scores"][var]["feedback"]`. Modify the `data` dictionary in-place.                                                 |
| `file()`     | :x:                | N/A. Returns an `object` (string, bytes-like, file-like)                                                 | Generate a file object dynamically in lieu of a physical file. Trigger via `type="dynamic"` in the question element (e.g., `pl-figure`, `pl-file-download`). Access the requested filename via `data['filename']`. If `file()` returns nothing, an empty string will be used. |

> As shown in the table, most functions accept a single argument, `data` (a dictionary), and modify it in place.

Reviewer 2: ```suggestion
As shown in the table, all functions (except for `render`) accept a single argument, `data` (a dictionary), and modify it in place. The `render` function accepts two arguments: the `data` dictionary and the `html` content computed from the template and elements.
```

Author: Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.

| `correct_answers`       | `dict`  | Correct answers for the question variant. Each item maps from a named answer to a value.                                             |
| `submitted_answers`     | `dict`  | Student answers submitted for the question after parsing.                                                                            |
| `raw_submitted_answers` | `dict`  | Raw student answers submitted for the question.                                                                                      |
| `format_errors`         | `dict`  | Dictionary of format errors for each answer. Each item maps from a named answer to a error message.                                  |
| `partial_scores`        | `dict`  | Dictionary of partial scores for each answer. Each entry is a dictionary with the keys `score` (float) and `weight` (int, optional). |
| `score`                 | `float` | The total score for the question variant.                                                                                            |

> | `feedback`              | `dict`  | Dictionary of feedback for each answer. Each item maps from a named answer to a feedback message.                                    |

Reviewer 2: Missing: `variant_seed` (all phases), `options` (all phases), `filename` (for `file()`). There are also a few specific to `render()` (`editable`, `manual_grading`, `panel`, `num_valid_submissions`), `prepare()` (`answers_name`) and `test()` (`test_type`), but I don't think they need to be listed since the docs below point to them.

Reviewer 2: To be clear: I don't think they need extensive documentation, but at least a link to where they are described may be useful. All three of them have a description in the same file.

Author: Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.

## Step 2: `prepare`

> Next, the `prepare` function is called after all elements (e.g. `<pl-integer-input>`) have run `prepare()`. This is may be done to do any sort of final post-processing, but is not commonly used.

Reviewer 2: ```suggestion
Next, the `prepare` function is called after all elements (e.g. `<pl-integer-input>`) have run `prepare()`. This may be done to do any sort of final post-processing, but is not commonly used.
```

        data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
```

## `server.py` functions

This table summarizes the functions that can be defined in `server.py`.

> | Function     | Updates `data`?    | Modifiable `data` keys                                                                                   | Description                                                                                                                                                                                                                                                                   |

Author: This table removes a column so that the entire table can be displayed without scrolling. I think this tradeoff is well worth it, and makes viewing this information easier at a glance.

| `grade()`    | :white_check_mark: | `correct_answers`, `feedback`, `format_errors`, `params`, `partial_scores`, `score`, `submitted_answers` | Grade `data["submitted_answers"][var]` to determine a score. Store the score and any feedback in `data["partial_scores"][var]["score"]` and `data["partial_scores"][var]["feedback"]`. Modify the `data` dictionary in-place.                                                 |
| `file()`     | :x:                | N/A. Returns an `object` (string, bytes-like, file-like)                                                 | Generate a file object dynamically in lieu of a physical file. Trigger via `type="dynamic"` in the question element (e.g., `pl-figure`, `pl-file-download`). Access the requested filename via `data['filename']`. If `file()` returns nothing, an empty string will be used. |

As shown in the table, most functions accept a single argument, `data` (a dictionary), and modify it in place.

### `QuestionData` dictionary

<!-- prettier-ignore -->

> ::: prairielearn.question_utils.QuestionData

Author: This has an ugly black line above it -- is it worth removing the black line `extra.css` styles from all documentation so that this looks better? It was intended as a separator for the reference documentation.

    options:
        heading_level: 3
        show_root_full_path: false
        show_root_toc_entry: false
        show_root_heading: false
        show_docstring_attributes: true

>         members: []

Author: This means that we will only show the `attributes` docstring, which only includes properties that are commonly set.

From within `server.py` functions, directories can be accessed as:

```python
# on-disk location of the current question directory
data["options"]["question_path"]

# on-disk location of clientFilesQuestion/
data["options"]["client_files_question_path"]

# URL location of clientFilesQuestion/ (only in render() function)
data["options"]["client_files_question_url"]

# URL location of dynamically-generated question files (only in render() function)
data["options"]["client_files_question_dynamic_url"]

# on-disk location of clientFilesCourse/
data["options"]["client_files_course_path"]

# URL location of clientFilesCourse/ (only in render() function)
data["options"]["client_files_course_url"]

# on-disk location of serverFilesCourse/
data["options"]["server_files_course_path"]
```

## Generating dynamic files with `file()`

> You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. They can also access `data["params"]`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:

Reviewer 3: ```suggestion
You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. `file()` has access to the full `data` object, including `data["params"]`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
```

From within `server.py` functions, directories can be accessed as:

```python

> # on-disk location of the current question directory

Author: Done to avoid scrolling right.

data["options"]["question_path"]

# on-disk location of clientFilesQuestion/
data["options"]["client_files_question_path"]

# URL location of clientFilesQuestion/ (only in render() function)
data["options"]["client_files_question_url"]

# URL location of dynamically-generated question files (only in render() function)
data["options"]["client_files_question_dynamic_url"]

# on-disk location of clientFilesCourse/
data["options"]["client_files_course_path"]

# URL location of clientFilesCourse/ (only in render() function)
data["options"]["client_files_course_url"]

# on-disk location of serverFilesCourse/
data["options"]["server_files_course_path"]
```

## Generating dynamic files with `file()`

> You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. These files are generated before the question is rendered in `question.html`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:

Author: The current lifecycle diagram on `index.md` doesn't include this step. Is it worth including?

Reviewer 3: It can't hurt!

Author: Added in [9b81bf8](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/9b81bf837e7c933a426590a9586f3c3b9bf27c5b).

From within `server.py` functions, directories can be accessed as:

```python
# on-disk location of the current question directory
data["options"]["question_path"]

# on-disk location of clientFilesQuestion/
data["options"]["client_files_question_path"]

# URL location of clientFilesQuestion/ (only in render() function)
data["options"]["client_files_question_url"]

# URL location of dynamically-generated question files (only in render() function)
data["options"]["client_files_question_dynamic_url"]

# on-disk location of clientFilesCourse/
data["options"]["client_files_course_path"]

# URL location of clientFilesCourse/ (only in render() function)
data["options"]["client_files_course_url"]

# on-disk location of serverFilesCourse/
data["options"]["server_files_course_path"]
```

## Generating dynamic files with `file()`

> You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. These files are generated before the question is rendered in `question.html`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:

Reviewer 3: > These files are generated before the question is rendered in `question.html`.

This is not true; did you get this impression from somewhere in the docs? If so, let's fix wherever that was to clear things up.

Author: This was an incorrect guess; can you suggest the correct text?

Author: This will help me inform https://github.com/PrairieLearn/PrairieLearn/pull/11830#discussion_r2054537142

Author: It needs to happen before the page is actually shown to the user, so I guess it happens after `render` but before it's actually rendered up? Or is this `file` function called dynamically upon requests?

Reviewer 3: > Or is this `file` function called dynamically upon HTML requests?

This is exactly correct. Breadcrumbs:

- https://github.com/PrairieLearn/PrairieLearn/blob/396f5428db5d517b7e37ae47dd021553439cbde5/apps/prairielearn/src/question-servers/freeform.js#L1233
- https://github.com/PrairieLearn/PrairieLearn/blob/330f53fb3ea27caf759e6a2ace840087ec7760a6/apps/prairielearn/src/lib/question-render.ts#L222
- https://github.com/PrairieLearn/PrairieLearn/blob/330f53fb3ea27caf759e6a2ace840087ec7760a6/apps/prairielearn/src/pages/generatedFilesQuestion/generatedFilesQuestion.ts#L44-L51

Author: Updated in [9b81bf8](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/9b81bf837e7c933a426590a9586f3c3b9bf27c5b).

    data["params"]["a"] = random.choice([0.25, 0.5, 1, 2, 4])

def file(data):
    # check for the appropriate filename
    if data["filename"] == "fig.png":
        # plot a line with slope "a"
        plt.plot([0, data["params"]["a"]], [0, 1])
        # make a bytes object (a buffer)
        buf = io.BytesIO()
        # save the figure data into the buffer
        plt.savefig(buf, format="png")
        return buf
```

> Note the use of `client_files_question_dynamic_url` in the `question.html` file. Requests for files to this URL will be routed to the `file()` function in `server.py`. The filename requested is stored in `data["filename"]`, and the file contents should be returned from the `file()` function. The URL for the dynamically generated files is set to `{{options.client_files_question_dynamic_url}}` in the HTML, which is a special URL that PrairieLearn uses to route requests to the `file()` function.

Reviewer 3: In general, we'd prefer it if people use `<pl-figure>` directly for this; can we update the example above, this paragraph, and the following one to reflect that? I'm happy to keep a reference to `client_files_question_dynamic_url` for advanced users, but that should indeed be an "advanced" alternative to elements that can handle this automatically.

Reviewer 2: > In general, we'd prefer it if people use `<pl-figure>` directly for this

Or `pl-file-download` for downloadable non-images.

> I'm happy to keep a reference to `client_files_question_dynamic_url` for advanced users, but that should indeed be an "advanced" alternative to elements that can handle this automatically.

That reference could point to examples of dynamic files that are neither images nor downloadable files. PDF embeds comes to mind as an example, but I don't know if you want to document that.

Author: I updated this section in [6ab669a](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/6ab669a078684c0733dd46b936ece12813b92b53)

Reviewer 3: > Or `pl-file-download` for downloadable non-images.

Yep; I was specifically talking about the example above where we were using `<img src="..." />`, which maps directly to `<pl-figure>`.

> PDF embeds comes to mind as an example, but I don't know if you want to document that.

I think PDF embeds are almost always an anti-pattern compared to just making them available for download.

Author: Small wording tweak in [8b77737](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/8b777377f6a58dd85585dd3b2bc2dcff255033b5).

Reviewer 3: Looks good!

    data["params"]["a"] = random.choice([0.25, 0.5, 1, 2, 4])

def file(data):
    # check for the appropriate filename
    if data["filename"] == "fig.png":
        # plot a line with slope "a"
        plt.plot([0, data["params"]["a"]], [0, 1])
        # make a bytes object (a buffer)
        buf = io.BytesIO()
        # save the figure data into the buffer
        plt.savefig(buf, format="png")
        return buf
```

> Requests for files to this URL will be routed to the `file()` function in `server.py` since the `type` of `pl-figure` is set to `"dynamic"`. The filename requested is stored in `data["filename"]`, and the file contents should be returned from the `file()` function. The URL for the dynamically generated files is set to `{{options.client_files_question_dynamic_url}}` in the HTML, which is a special URL that PrairieLearn uses to route requests to the `file()` function.

Reviewer 3: Can we move the nmote about `{{options...}}` to the "advanced usage" section below?

Author: Moved in [180b6d8](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/180b6d841e4202cfcceadd30f94c48524a62b60c).

> We recommend using the [`pl-figure`](../elements.md#pl-figure-element) and [`pl-file-download`](../elements.md#pl-file-download-element) elements to display or download files. Specifically, you should use `pl-file-download` for PDFs and other content that should be downloaded, and `pl-figure` for images and other content that should be displayed.

Reviewer 2: ```suggestion
We recommend using the [`pl-figure`](../elements.md#pl-figure-element) and [`pl-file-download`](../elements.md#pl-file-download-element) elements to display or download files. Specifically, you should use `pl-file-download` for files that should be downloaded or shown in a separate tab (e.g., PDFs, source code, etc.), and `pl-figure` for images that should be displayed.
```

Author: I made this change in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9.

### Recalculating scores

Any custom grading function for the whole question should set `data["score"]` as a value between 0.0 and 1.0, which will be the final score for the given question. If a custom grading function is only grading a specific part of a question, the grading function should set the corresponding dictionary entry in `data["partial_scores"]` and then recompute the final `data["score"]` value for the whole question. The `prairielearn` Python library provides the following score recomputation functions:

- [`set_weighted_score_data`][prairielearn.question_utils.set_weighted_score_data] sets `data["score"]` to be the weighted average of entries in `data["partial_scores"]`.
- [`set_all_or_nothing_score_data`][prairielearn.question_utils.set_all_or_nothing_score_data] sets `data["score"]` to 1.0 if all entries in `data["partial_scores"]` are 1.0, 0.0 otherwise.

This can be used like so:

```python title="server.py"
from prairielearn import set_weighted_score_data

def grade(data):
    # update partial_scores as necessary
    # ...

    # compute total question score

>     set_weighted_score_data(data)

Reviewer 3: Nit: if this is a best practice, can we update the example above with this?

Author: 8de961fa8bacdf280e06dd8f5e73d124836d5b15

```

> More detailed information can be found in the [grading utilities documentation](http://localhost:8000/python-reference/prairielearn/grading_utils/). If you prefer not to show score badges for individual parts, you may unset the dictionary entries in `data["partial_scores"]` once `data["score"]` has been computed.

Reviewer 3: You probably don't want a reference to `http://localhost:8000` in here?

Author: 8de961fa8bacdf280e06dd8f5e73d124836d5b15

From within `server.py` functions, directories can be accessed as:

```python
# on-disk location of the current question directory
data["options"]["question_path"]

# on-disk location of clientFilesQuestion/
data["options"]["client_files_question_path"]

# URL location of clientFilesQuestion/ (only in render() function)
data["options"]["client_files_question_url"]

# URL location of dynamically-generated question files (only in render() function)
data["options"]["client_files_question_dynamic_url"]

# on-disk location of clientFilesCourse/
data["options"]["client_files_course_path"]

# URL location of clientFilesCourse/ (only in render() function)
data["options"]["client_files_course_url"]

# on-disk location of serverFilesCourse/
data["options"]["server_files_course_path"]
```

## Generating dynamic files with `file()`

> You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. `file()` has access to the full `data` object, including `data["params"]`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:

Reviewer 2: `file()` does not have access to the whole object, AFAIK it does not have access to submitted answers, format errors, feedback, etc.

```suggestion
You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. `file()` has access to the same `data` object as the one set by the `generate()` function, including `data["params"]` and `data["correct_answers"]`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
```

Author: Changed in 6408ede3ac9fe51ab8d193ea476887f9d181752a.
